File: android/app/src/main/kotlin/io/dotnetnativeinterop/transport/pb/PqHandshakeClient.kt
package io.dotnetnativeinterop.transport.pb

import com.google.protobuf.ByteString
import dni.frame.v1.Envelope
import dni.frame.v1.HandshakeReply
import org.bouncycastle.crypto.kems.MLKEMGenerator
import org.bouncycastle.crypto.params.MLDSAParameters
import org.bouncycastle.crypto.params.MLDSAPublicKeyParameters
import org.bouncycastle.crypto.params.MLKEMParameters
import org.bouncycastle.crypto.params.MLKEMPublicKeyParameters
import org.bouncycastle.crypto.signers.MLDSASigner
import java.security.SecureRandom

/** The negotiated params of a completed PQ handshake — feeds the Trust inspector's live-params card. */
internal data class PqChannelParams(
    val kem: String,
    val sig: String,
    val cipher: String,
    val kemPublicKeyBytes: Int,
    val ciphertextBytes: Int,
    val sharedSecretBytes: Int,
    val handshakeUs: Double,
)

/** The outcome of a client handshake: the per-frame cipher plus the negotiated params for the UI. */
internal class PqHandshakeResult(val cipher: AeadFrameCipher, val params: PqChannelParams)

/** Thrown when the PQ handshake cannot be completed (bad offer, failed signature, wrong frame type). */
internal class PqHandshakeException(message: String) : Exception(message)

/**
 * Client side of the framed-protobuf PQ handshake — the mirror of the engine's
 * `PqSecureChannel.ServerHandshakeAsync`, run over a still-plaintext [FrameChannel] before any
 * application frames flow. One round trip:
 *
 *   server -> client  HandshakeOffer  { ML-KEM-768 public key, ML-DSA-65 public key, session_id (fresh),
 *                                        ML-DSA signature over (kem_public_key || session_id), alg labels }
 *   client -> server  HandshakeReply  { ML-KEM ciphertext encapsulating the shared secret }
 *
 * The client verifies the signature over (kem_public_key || session_id) BEFORE encapsulating, then both
 * sides HKDF-SHA256 the shared secret into two directional AES-256-GCM keys (salt = ciphertext ||
 * session_id, so a replayed ciphertext can never resurrect the same keys) and switch the channel to
 * [AeadFrameCipher]. Uses BouncyCastle's LIGHTWEIGHT `org.bouncycastle.crypto.*` API directly — never
 * the JCA "BC" provider, whose Android-bundled legacy copy lacks these PQC classes entirely.
 *
 * SECURITY NOTE (loopback teaching artifact): the ML-DSA signature authenticates handshake transcript
 * INTEGRITY against a passive tamperer — it does NOT pin the server's identity (trust-on-first-use).
 * Nothing here checks that the offered sig_public_key belongs to the expected server. Do not use off
 * loopback without pinning the server's ML-DSA public key.
 */
internal object PqHandshakeClient {
    // Directional HKDF info labels — ASCII, byte-identical to the engine's InfoClientToServer /
    // InfoServerToClient. The client SENDS on c2s and RECEIVES on s2c (mirror of the server).
    private val INFO_CLIENT_TO_SERVER = "dni-pb c2s v1".toByteArray(Charsets.US_ASCII)
    private val INFO_SERVER_TO_CLIENT = "dni-pb s2c v1".toByteArray(Charsets.US_ASCII)
    private const val KEY_BYTES = 32

    private val random = SecureRandom()

    /**
     * Runs the handshake over [channel] (which must NOT yet have a cipher attached) and returns the
     * negotiated cipher + params. Does NOT attach the cipher to the channel — the caller does that after
     * this returns, so the handshake frames themselves stay plaintext (as the server expects).
     */
    fun handshake(channel: FrameChannel): PqHandshakeResult {
        val start = System.nanoTime()

        val offerBytes = channel.readFrame()
            ?: throw PqHandshakeException("server closed the connection before sending a handshake offer")
        val offerEnvelope = Envelope.parseFrom(offerBytes)
        if (offerEnvelope.bodyCase != Envelope.BodyCase.HANDSHAKE_OFFER) {
            throw PqHandshakeException("expected a handshake offer but received ${offerEnvelope.bodyCase}")
        }

        val offer = offerEnvelope.handshakeOffer
        val kemPublicKey = offer.kemPublicKey.toByteArray()
        val sessionId = offer.sessionId.toByteArray()

        // Verify the ML-DSA signature over (kem_public_key || session_id) — exactly what the server signed.
        val message = concat(kemPublicKey, sessionId)
        if (!verify(offer.sigPublicKey.toByteArray(), message, offer.signature.toByteArray())) {
            throw PqHandshakeException("server handshake signature failed ML-DSA verification")
        }

        // Encapsulate to the server's ML-KEM public key -> (ciphertext on the wire, shared secret local).
        // A wrong-length/garbage key throws a raw BC IllegalArgumentException/DataLengthException; wrap it
        // so a malformed offer surfaces as the PqHandshakeException the class contract promises.
        val kemPub = wrapBc("invalid ML-KEM public key in offer") {
            MLKEMPublicKeyParameters(MLKEMParameters.ml_kem_768, kemPublicKey)
        }
        val encapsulated = wrapBc("ML-KEM encapsulation failed") { MLKEMGenerator(random).generateEncapsulated(kemPub) }
        val ciphertext = encapsulated.encapsulation
        val sharedSecret = encapsulated.secret

        val reply = Envelope.newBuilder()
            .setHandshakeReply(HandshakeReply.newBuilder().setCiphertext(ByteString.copyFrom(ciphertext)))
            .build()
        channel.writeFrame(reply.toByteArray())

        // Derive the two directional keys. salt = ciphertext || session_id (session_id in the SALT, not the
        // info, so a fresh session_id randomizes the extracted PRK — matching the engine's DeriveKeys).
        val salt = concat(ciphertext, sessionId)
        val clientToServer = Hkdf.deriveKey(sharedSecret, KEY_BYTES, salt, INFO_CLIENT_TO_SERVER)
        val serverToClient = Hkdf.deriveKey(sharedSecret, KEY_BYTES, salt, INFO_SERVER_TO_CLIENT)
        val cipher = AeadFrameCipher(sendKey = clientToServer, recvKey = serverToClient)

        val handshakeUs = (System.nanoTime() - start) / 1000.0
        val params = PqChannelParams(
            kem = offer.kemAlgorithm.ifEmpty { "ML-KEM-768" },
            sig = offer.sigAlgorithm.ifEmpty { "ML-DSA-65" },
            cipher = offer.cipher.ifEmpty { AeadFrameCipher.ALGORITHM },
            kemPublicKeyBytes = kemPublicKey.size,
            ciphertextBytes = ciphertext.size,
            sharedSecretBytes = sharedSecret.size,
            handshakeUs = handshakeUs,
        )
        return PqHandshakeResult(cipher, params)
    }

    // Verifies an ML-DSA-65 signature over [message] with the server's encoded public key. Private: only
    // handshake() uses it. A malformed public key/signature (raw BC exception) is wrapped as a
    // PqHandshakeException; a well-formed-but-wrong signature simply returns false.
    private fun verify(sigPublicKey: ByteArray, message: ByteArray, signature: ByteArray): Boolean =
        wrapBc("ML-DSA verification error") {
            val publicKey = MLDSAPublicKeyParameters(MLDSAParameters.ml_dsa_65, sigPublicKey)
            val verifier = MLDSASigner()
            verifier.init(false, publicKey)
            verifier.update(message, 0, message.size)
            verifier.verifySignature(signature)
        }

    // Runs a BouncyCastle call, translating its raw RuntimeExceptions (IllegalArgumentException,
    // DataLengthException, …) into the PqHandshakeException the class contract promises. A
    // PqHandshakeException already thrown inside passes through unchanged.
    private inline fun <T> wrapBc(what: String, block: () -> T): T =
        try {
            block()
        } catch (e: PqHandshakeException) {
            throw e
        } catch (e: RuntimeException) {
            throw PqHandshakeException("$what: ${e.message}")
        }

    private fun concat(first: ByteArray, second: ByteArray): ByteArray {
        val result = ByteArray(first.size + second.size)
        System.arraycopy(first, 0, result, 0, first.size)
        System.arraycopy(second, 0, result, first.size, second.size)
        return result
    }
}

File: ios/Shared/Pb/PqHandshakeClient.swift
import CryptoKit
import Foundation

/// The negotiated params of a completed PQ handshake — feeds the Trust inspector's live-params card.
/// Mirrors the engine's / Android's `PqChannelParams`.
struct PqChannelParams: Sendable, Equatable {
    let kem: String
    let sig: String
    let cipher: String
    let kemPublicKeyBytes: Int
    let ciphertextBytes: Int
    let sharedSecretBytes: Int
    let handshakeUs: Double
}

/// The outcome of a client handshake: the per-frame cipher plus the negotiated params for the UI.
/// Gated with the cipher (iOS 26+); `PqChannelParams` above is plain data and stays always-available.
@available(iOS 26.0, macOS 15.0, *)
struct PqHandshakeResult {
    let cipher: AeadFrameCipher
    let params: PqChannelParams
}

enum PqHandshakeError: LocalizedError {
    case noOffer                    // peer closed before sending the offer
    case wrongFrame(String)         // a frame that isn't a HandshakeOffer
    case badKemKey(String)          // malformed ML-KEM public key in the offer
    case signatureFailed            // ML-DSA verification of the offer failed (fatal)

    var errorDescription: String? {
        switch self {
        case .noOffer:            return "framed-protobuf PQ: server closed before sending a handshake offer"
        case .wrongFrame(let b):  return "framed-protobuf PQ: expected a handshake offer but got \(b)"
        case .badKemKey(let m):   return "framed-protobuf PQ: invalid ML-KEM public key in offer (\(m))"
        case .signatureFailed:    return "framed-protobuf PQ: server handshake signature failed ML-DSA verification"
        }
    }
}

/// Client side of the framed-protobuf PQ handshake — the CryptoKit mirror of the engine's
/// `PqSecureChannel.ServerHandshakeAsync` and Android's `PqHandshakeClient`, run over a still-plaintext
/// `PbConnection` before any application frames flow. One round trip:
///
///   server -> client  HandshakeOffer  { ML-KEM-768 public key, ML-DSA-65 public key, session_id (fresh),
///                                        ML-DSA signature over (kem_public_key || session_id), alg labels }
///   client -> server  HandshakeReply  { ML-KEM ciphertext encapsulating the shared secret }
///
/// The client verifies the signature over (kem_public_key || session_id) BEFORE encapsulating, then both
/// sides HKDF-SHA256 the shared secret into two directional AES-256-GCM keys (salt = ciphertext ||
/// session_id, so a replayed ciphertext can never resurrect the same keys) and switch to `AeadFrameCipher`.
/// CryptoKit implements the SAME FIPS-final parameter sets (ML-KEM-768 / ML-DSA-65) as the .NET/Android
/// BouncyCastle provider, so the shared secret — and every derived key — is byte-compatible cross-platform.
///
/// SECURITY NOTE (loopback teaching artifact): the ML-DSA signature authenticates handshake transcript
/// INTEGRITY against a passive tamperer — it does NOT pin the server's identity (trust-on-first-use).
/// Do not use off loopback without pinning the server's ML-DSA public key.
@available(iOS 26.0, macOS 15.0, *)
enum PqHandshakeClient {
    // Directional HKDF info labels — ASCII, byte-identical to the engine's / Android's. The client SENDS
    // on c2s and RECEIVES on s2c (mirror of the server).
    private static let infoClientToServer = Data("dni-pb c2s v1".utf8)
    private static let infoServerToClient = Data("dni-pb s2c v1".utf8)
    private static let keyBytes = 32

    /// Runs the handshake over `connection` (which must NOT yet have a cipher attached) and returns the
    /// negotiated cipher + params. The caller attaches the cipher after this returns, so the handshake
    /// frames themselves stay plaintext (as the server expects). `nowNs` supplies the monotonic clock so
    /// the caller can time the handshake without this depending on a global clock.
    static func handshake(_ connection: PbConnection, nowNs: () -> UInt64) throws -> PqHandshakeResult {
        let start = nowNs()

        guard let offerEnvelope = try connection.readEnvelope() else { throw PqHandshakeError.noOffer }
        guard case .handshakeOffer(let offer)? = offerEnvelope.body else {
            throw PqHandshakeError.wrongFrame(String(describing: offerEnvelope.body))
        }

        let kemPublicKey = offer.kemPublicKey
        let sessionId = offer.sessionID

        // Verify the ML-DSA-65 signature over (kem_public_key || session_id) — exactly what the server
        // signed. A malformed key/sig, or a well-formed-but-wrong signature, is fatal.
        let signedMessage = kemPublicKey + sessionId
        guard let sigPub = try? MLDSA65.PublicKey(rawRepresentation: offer.sigPublicKey),
              sigPub.isValidSignature(offer.signature, for: signedMessage) else {
            throw PqHandshakeError.signatureFailed
        }

        // Encapsulate to the server's ML-KEM public key -> (ciphertext on the wire, shared secret local).
        let kemPub: MLKEM768.PublicKey
        do {
            kemPub = try MLKEM768.PublicKey(rawRepresentation: kemPublicKey)
        } catch {
            throw PqHandshakeError.badKemKey(error.localizedDescription)
        }
        let encapsulated = try kemPub.encapsulate()
        let ciphertext = encapsulated.encapsulated
        let sharedSecret = encapsulated.sharedSecret          // SymmetricKey (the 32-byte KEM secret)

        var reply = Dni_Frame_V1_Envelope()
        var handshakeReply = Dni_Frame_V1_HandshakeReply()
        handshakeReply.ciphertext = ciphertext
        reply.handshakeReply = handshakeReply
        try connection.writeRequest(reply)                    // still plaintext (cipher not yet attached)

        // Derive the two directional keys. salt = ciphertext || session_id (session_id in the SALT, so a
        // fresh session_id randomizes the extracted PRK — matching the engine's / Android's DeriveKeys).
        let salt = ciphertext + sessionId
        let clientToServer = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: sharedSecret, salt: salt, info: infoClientToServer, outputByteCount: keyBytes)
        let serverToClient = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: sharedSecret, salt: salt, info: infoServerToClient, outputByteCount: keyBytes)
        let cipher = AeadFrameCipher(sendKey: clientToServer, recvKey: serverToClient)

        let handshakeUs = Double(nowNs() - start) / 1000.0
        let params = PqChannelParams(
            kem: offer.kemAlgorithm.isEmpty ? "ML-KEM-768" : offer.kemAlgorithm,
            sig: offer.sigAlgorithm.isEmpty ? "ML-DSA-65" : offer.sigAlgorithm,
            cipher: offer.cipher.isEmpty ? AeadFrameCipher.algorithm : offer.cipher,
            kemPublicKeyBytes: kemPublicKey.count,
            ciphertextBytes: ciphertext.count,
            sharedSecretBytes: sharedSecret.withUnsafeBytes { $0.count },
            handshakeUs: handshakeUs)
        return PqHandshakeResult(cipher: cipher, params: params)
    }
}
